x/ccv/provider/keeper/keeper_test.go

```diff
@@ -277,17 +271,17 @@ func (suite *KeeperTestSuite) TestInitHeight() {
 		{expected: 12, chainID: "chain2"},
 	}
 
-	app.ProviderKeeper.SetInitChainHeight(ctx, tc[1].chainID, tc[1].expected)
-	app.ProviderKeeper.SetInitChainHeight(ctx, tc[2].chainID, tc[2].expected)
+	providerKeeper.SetInitChainHeight(ctx, tc[1].chainID, tc[1].expected)
+	providerKeeper.SetInitChainHeight(ctx, tc[2].chainID, tc[2].expected)
 
-	for _, t := range tc {
-		height := app.ProviderKeeper.GetInitChainHeight(ctx, t.chainID)
-		suite.Require().EqualValues(t.expected, height)
+	for _, tc := range tc {
+		height := providerKeeper.GetInitChainHeight(ctx, tc.chainID)
+		require.Equal(t, tc.expected, height)
 	}
 }
 
 func (suite *KeeperTestSuite) TestHandleSlashPacketDoubleSigning() {
-	ProviderKeeper := suite.providerChain.App.(*appProvider.App).ProviderKeeper
+	providerKeeper := suite.providerChain.App.(*appProvider.App).ProviderKeeper
```

Comment:
Why not use the `GetProviderKeeperAndCtx` function? Similarly, why do we need to use `suite.ctx` in the slashing tests?

Reply:
Valid question, in this particular case I was able to make an equivalent test function using native golang testing and testify mocking (see [new test](https://github.com/cosmos/interchain-security/blob/56d61919c9c3d8cd548798f0e3205ee731aaac70/x/ccv/provider/keeper/keeper_test.go#L327)). However, I think it'd be best to still keep the e2e test version, as the e2e test validates the interaction between keepers at a higher level.

Looking at the bigger picture, the decision of unit tests vs e2e tests will have to be made on a case by case basis. In order to test this specific 
```TestHandleSlashPacketDoubleSigning``` functionality with [mock](https://pkg.go.dev/github.com/stretchr/testify/mock), there was a decent amount of [function call expectations I had to specify](https://github.com/cosmos/interchain-security/blob/56d61919c9c3d8cd548798f0e3205ee731aaac70/x/ccv/provider/keeper/keeper_test.go#L342). But I was able to test the slash packet functionality at a more granular level.

Reply:
That makes sense. I think it is useful to have both -- e2e tests that are checking that e.g. consumer initiated slashing works and unit tests that (as you said) are checking functionalities at a more granular level.